metaflow/plugins/cards/card_decorator.py

```diff
@@ -102,6 +153,51 @@ def task_pre_step(
         ubf_context,
         inputs,
     ):
+        # We have a step counter to ensure that on calling the final card decorator's `task_pre_step`
+        # we call a `finalize` function in the `CardComponentCollector`.
+        # This can help ensure the behaviour of the `current.card` object is according to specification.
+        self._increment_step_counter()
+        self._user_set_card_id = self.attributes["id"]
+        if (
+            self.attributes["id"] is not None
+            and re.match(CARD_ID_PATTERN, self.attributes["id"]) is None
+        ):
+            # There should be a warning issued to the user that `id` doesn't match regex pattern
+            # Since it is doesn't match pattern, we need to ensure that `id` is not accepted by `current`
+            # and warn users that they cannot use id for thier arguements.
+            wrn_msg = (
+                "@card with id '%s' doesn't match REGEX pattern. "
+                "Adding custom components to cards will not be accessible via `current.card['%s']`. "
+                "Please create `id` of pattern %s. "
+            ) % (self.attributes["id"], self.attributes["id"], TYPE_CHECK_REGEX)
+            warning_message(wrn_msg, self._logger)
+            self._user_set_card_id = None
+
+        # As we have multiple decorators,
+        # we need to ensure that `current.card` has `CardComponentCollector` instantiated only once.
+        if not self._is_event_registered("pre-step"):
```

Comment:
This feels heavy handed for effectively a single bool (it's not used elsewhere unless I am missing something).